Result recorder

kyleknap opened the pull request:
The abstraction tracks statistic based on ``Results`` passed to it

It currently tracks the following:
* Expected number of files that will be transferred 
* Expected number of bytes that will be transferred
* Number of bytes transferred so far
* Number of files transferred so far
* Number of warnings received so far
* Number of bytes failed to be transferred so far

I was not sure if it made sense to track number of bytes failed to be transferred so far and considered just combining it with number of bytes transferred, but that seemed a little weird. In the end I will need to account for that number when printing progress. So I am up for any suggestions.

I am also up for any other statistics to track, but I was just trying to keep it small for now.

cc @jamesls @JordonPhillips

jamesls commented on awscli/customizations/s3/results.py:
```diff
@@ -121,3 +122,77 @@ def _get_src_dest(self, future):
+            result=result)
+
+    def _get_progress_cache_key(self, result):
+        return ':'.join([result.transfer_type, result.src, result.dest])
```
Do you think it's an issue that this code relies on the property that any key in `_result_handler_map` must have a `transfer_type`, `src`, and `dest`.  There's not really a base type that we can say all `result` params must subclass from `X`.

I think there's potential that if we add more result types we might forget this requirement.

kyleknap replied:
@jamesls
Yeah possibly. I can add a ``SubscriberResult`` base class that makes the three params required for any results produced from a ``ResultSubscriber``. Yet again I am not sure how many more result types that I will need to add. Right now I can really only see it being needed for dryrun results maybe (but that may require a different handler)? Otherwise it should not be required for all Results as stuff like ``WarningResults`` only requires a message. What do you think? I would be more on the side if I find myself adding any more result classes that require this, then add it?

jamesls replied:
That seems reasonable.  Even if a type check was added, it would still likely be an error case anyways, and the main benefit there would just be a better error message than `AttributeError`.

kyleknap replied:
Yeah I like the type check idea. Just a quick helpful reminder for the future. I will add that.

jamesls commented:
Overall I think it looks good.

Do all the callbacks invoked on the `ResultRecorder` object happen in the same thread?  As long as that's the case I don't think there's any issues.  Otherwise there might be some race conditions to handle.

kyleknap commented:
Yep the ResultRecorder will be used in only one thread so should not have to worry about concurrency issues.

jamesls commented:
:shipit:

kyleknap commented:
@jamesls I just updated to subclass from a ``BaseResult`` class and added the type check. Let me know what you think about how I did it.

I was not sure what tests to add for the error case as it is more of a help case for us when we are developing and it is impossible to replicate unless I inject new custom behavior into the ResultRecorder, which would be weird for a test.

jamesls commented on awscli/customizations/s3/results.py:
```diff
@@ -21,21 +21,31 @@
+
+BaseResult = namedtuple('BaseResult', ['transfer_type', 'src', 'dest'])
+
+QueuedResult = _create_new_result_cls('QueuedResult', ['total_transfer_size'])
```
It's a little weird to have `_create_new_result_cls` implicitly depend on `BaseResult` which is defined above.  If the `BaseResult` line is moved below a `_create_new_result_cls` the function will fail.  What do you think about passing `BaseResult` to `_create_new_result_cls` to make this more explicit?

kyleknap replied:
Makes sense I can add that. Would make it more extensible as well.

kyleknap commented:
@jamesls I updated the result class factory by allowing you to inject a base result class. Should be good to look at again.

jamesls commented:
:shipit:
